feat: ✨  add commit guidelines (#58)
seedcase-project/guidebook: workflows/commits.qmd

# Commits

This guide covers how we in the Seedcase Project commit changes to our
repositories. We follow a set of guidelines to ensure that our commit

> history is clear and consistent which allow us to automatically

Reviewer: ```suggestion
history is clear and consistent, which allows us to automatically
```
if it's about the state of the commit history

generate changelogs and version updates.

Below, we'll go through when and what to commit, the Conventional
Commits specification, and how to apply it to both code and non-code
projects.

## When and what to commit

In the Seedcase Project, we follow a few guidelines for when and what we
commit. We aim to:

> -   Make atomic commits: Each commit should represent a single logical

Reviewer 2: ```suggestion
-   Make atomic commits/pull requests: Each commit (or pull request) should represent a single logical
```

Commits](https://marketplace.visualstudio.com/items?itemName=vivaxy.vscode-conventional-commits)
to help us write conventional commit messages. It allows us to add
emojis following the [Gitmoji](https://gitmoji.dev/) specification to
make the messages more expressive and visually appealing.

So, if you use VS Code, we recommend installing this extension to ease
the process of writing conventional commit messages.
:::

### Required: type and description

The `<type>` in Conventional Commits is inspired by [Angular's commit
message
conventions](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#-commit-message-guidelines)
and is used to categorise the changes made in the commit:

-   `build`: Changes that affect the build system or external
    dependencies.
-   `ci`: Changes to our CI configuration files and scripts.
-   `docs`: Documentation only changes.
-   `feat`: A new feature.
-   `fix`: A bug fix.
-   `perf`: A code change that improves performance.
-   `refactor`: A code change that neither fixes a bug nor adds a
    feature.
-   `style`: Changes that do not affect the meaning of the code
    (white-space, formatting, missing semi-colons, etc).

> -   `test`: Adding missing tests or correcting existing tests.

Reviewer 2: ```suggestion
-   `test`: Adding tests or correcting existing tests.
```

-   `revert`: Reverts a previous commit.

> -   `chore`: Other changes that don't modify src or test files.

Reviewer 2: ```suggestion
-   `chore`: Other changes that don't modify source or test files.
```

The `<description>` is a short summary of the change. It should be
descriptive enough to understand the change without needing to look at
actual changes. It is written in the imperative, present tense
(i.e., "change" not "changed" nor "changes"), is not capitalized, and is
without a period at the end.

### Optional: scope, body, and footer(s)

In the Seedcase Project, we tend not to use `[optional scope]` that

> often, mostly for automatic commits that are generated by tools like

Reviewer: ```suggestion
often, mostly only for automatic commits that are generated by tools like
```
Or something to undo the negative in the first clause

Author: ```suggestion
often. We mostly use it in automatic commits that are generated by tools like
```

[Dependabot](https://github.com/dependabot).

Whenever we want to expand on the `<description>`, we add an
`[optional body]` that provides more context or details about the
change. This can include the motivation for the change, how it was
implemented, or any other relevant information that didn't fit in the
description.

The `[optional footer(s)]` can be used for various purposes, such as
indicating breaking changes or referencing issues. We rarely use them,
but if we want to indicate a breaking change, we will add a footer in
the format `BREAKING CHANGE: <description>`. Following the Conventional
Commits specification, whenever adding a breaking change, we will also
add a `!` after the `<type>` in the commit message. For example:

``` md
feat!: new feature that breaks existing functionality

BREAKING CHANGE: This feature changes the way the existing functionality
works, so it is not backward compatible.
```

## Conventional commits for non-code projects

The conventional commit specification was created for software or
code-related projects, but we also use it for documentation, workshops,

> and other non-code projects. In those cases, we adapt the use of

Reviewer: ```suggestion
and other non-code projects. In these cases, we adapt the use of
```

    the meaning or intent stays the same. This is used in editing or
    proofreading phases or tasks.
-   `fix`: Changes that fix typos, grammatical errors, incomplete
    sentences, or other similar writing mix-ups/situations where the
    text doesn't align with our original intent.

Each time we finish running a workshop, we’ll mark it with a
`BREAKING CHANGE: <description>` footer. This breaking change commit
will be made *soon after* the workshop has been run to include all
changes made right before and during the workshop. By including this
footer, we clearly indicate that this is a new version of the workshop
related to that specific iteration of it. This helps us track workshop
versions and the updates made to each one. Following the Conventional
Commits specification, whenever we add a breaking change, we will also
add a `!` after the `<type>` in the commit message to indicate that it
is a breaking change.

As for emojis, we generally use them in the same way as in software
projects. The only emoji exception is the `:memo:` (:memo:) emoji that
is used for documentation-related commits. In the same way as the `docs`

> type, it is redundant when almost all changes are related to

Reviewer 2: ```suggestion
`<type>`, it is redundant when almost all changes are related to
```

> documentation. Instead, we use other emojis that are more descriptive of

Reviewer 2: ```suggestion
documentation. Instead, use other emojis that are more descriptive of
```
